sky/backends/cloud_vm_ray_backend.py

```diff
@@ -2865,12 +2866,37 @@ def _sync_file_mounts(
         self._execute_file_mounts(handle, all_file_mounts)
         self._execute_storage_mounts(handle, storage_mounts)
 
+    def _update_envs_for_k8s(self, handle: CloudVmRayResourceHandle,
+                             task: task_lib.Task) -> None:
+        """Update envs with env vars from Kubernetes if cloud is Kubernetes.
+
+        Kubernetes automatically populates containers with critical environment
+        variables, such as those for discovering services running in the
+        cluster and CUDA/nvidia environment variables. We need to update task
+        environment variables with these env vars. This is needed for GPU
+        support and service discovery.
+
+        See https://github.com/skypilot-org/skypilot/issues/2287 for
+        more details.
+        """
+        if isinstance(handle.launched_resources.cloud, clouds.Kubernetes):
+            temp_envs = copy.deepcopy(task.envs)
+            cloud_env_vars = handle.launched_resources.cloud.query_env_vars(
+                handle.cluster_name)
+            task.update_envs(cloud_env_vars)
+
+            # Re update the envs with the original envs to give priority to
+            # the original envs.
+            task.update_envs(temp_envs)
```

Comment:
I would prefer to include those necessary env vars in the `~/.bashrc` instead of the users' task env vars. 
Reason: These env vars seem to be global for all the tasks submitted to the same VM, and for the user who would like to `ssh` into the VM, the `nvidia-smi` will not work if we only enable the env vars for the tasks.

Reply:
That's a good point! Accessing SSH vars is being worked on by @hemildesai in https://github.com/landscapepainter/skypilot/pull/1, and we plan to ship it in #2435.  If you think this is urgent, I can to add it `~/.bashrc` here as a stopgap fix and remove it in #2435.

------
More context - 

The challenge we are trying to solve here is that environment variables from Kubernetes are dynamic and change over time. 

One example, as #2287 mentions, is service discovery in Kubernetes. Kubernetes populates service IPs and ports as environment variables. These IPs and ports may change or new ones may be added, which requires us to query environment variables dynamically every time.

For SSH too, we want to query environment variables dynamically and populate the environment. In his PR, @hemildesai is doing so by adding a ProxyCommand which query envvars whenever a SSH connection is initiated and updates the environment variables in `/etc/environment`.

Reply:
Thanks for the detailed explanation!

One question I have is that, if the env vars are dynamic and the user ssh into the VM, and try to access the env var after a while, does that mean the env vars can be staled?

I think it should be fine to let the k8s related variables to be added in the future. One thing I think that is important is to make sure the `nvidia-smi` works through ssh, as it might be very confusing that `ssh cluster-name` and `nvidia-smi` fail to work, even though the VM is launched with GPU resources.

I am guessing this may involve some nvidia-smi related PATH to be exported?

Reply:
Totally agree - we should make `nvidia-smi` work for `ssh cluster-name`. This is fixed by #2435, and we plan to merge that PR also soon before we start advertising Kubernetes support, so I'll leave it out from this PR :) I have filed #2453 to make sure it gets fixed. Would that be ok?

> One question I have is that, if the env vars are dynamic and the user ssh into the VM, and try to access the env var after a while, does that mean the env vars can be staled?

Yes (though very rare), some envvars can be stale (e.g., if the service port gets changed by the user). These semantics work for Kubernetes applications because the pod will be automatically restarted if it crashes due to bad envvars, and the restarted pods will fetch latest envvars.

Reply:
Sounds good! Thanks for the explanation : )